### ISSUE TITLE
Port just enough of trial to run --version on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7803]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#7803|
|Type|enhancement|
|Created|2015-03-07 19:55:38Z|
|Branch|https://github.com/twisted/twisted/tree/py3-trial-version-7803-3|

Porting just enough to run --version lays out some groundwork that lets further changes focus on porting functionality, not just getting it importing at all.

Original desc:

This is done in parallel with [#5966](https://github.com/twisted/twisted/issues/5966) ... for which I see that there was no recent activity and that some proposed changes are already in trunk

My goal is to have an initial version of trial running on Python3 so that it can replace the run-python3-tests and later run its own tests suite and other tests.

I already have a local branch on which trial works on python3 with just a few changes to filepath and plugins registration using @implementer decorator... and it can run the tests executed by run-python3-tests 

The plan is to break my local branch into simple tickets and send them for review.

<details><summary>Searchable metadata</summary>

```
trac-id__7803 7803
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_py3_trial_version_7803_3 branches/py3-trial-version-7803-3
branch_author__hawkowl__adiroiban hawkowl, adiroiban
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1425758138104328 1425758138104328
changetime__1432186372519245 1432186372519245
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

(In [43986]) Branching to py3-trial-version-7803.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban removed owner|
|-|-|

I see that [#3843](https://github.com/twisted/twisted/issues/3843) ask to deprecate twisted.persisted but [#6910](https://github.com/twisted/twisted/issues/6910) and [#6911](https://github.com/twisted/twisted/issues/6911) were created to port them to Python3.

I have delayed the imports for twisted.persisted.sod for
runtime... in this as long as python3 apps don't use persisted all should be fine.

I had no idea how to port _preample to Pyhon3 so for now
I have created a trial3 script and in setup3.py installed **all** scripts. Please suggest a way to handle this.

I assume that spewer function is only used for testing and not in production... I check the code in spewer and it looks like it was not designed to be easy to write an automated test for it.
I tested it my branch which has a functional trial on py3 and it works.


I would like to have trial working on Python3 so that on my local
computer I can just run this even if trial is not 100% ported... but it can be used to port trial itself.

```
tox -e py26,py27,py33 twisted/application/test/test_internet.py
```


Please take a look at the changes and let me know what do you think.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi!

Few things:

- trial3 should have "#!/usr/bin/env python3", not "env python", as that's python2
- The comments about sob are a bit wonky, should be:

# twisted.persisted is proposed for deprecation and is not yet ported to
# to Python3 so we only import it if some code really needs to use it

- We don't want to install all scripts yet. bin/trial3 is not installed, but runs from the bin/ directory for developers
- Preamble doesn't need to be ported, it works fine on Py2 and Py3 (it seems)

Thanks! :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review!

I have removed trial3 as trial2 also works on Python3... I was confused by the _preamble import and this is why I have created trial3 script.

-------

I have updated the comments for twisted.persisted

-------

I have reverted the changes in dist3.py

-----

I have added a test for changes from t.application.services... please check test is ok.
I don't know what docstring to add at that test without duplicating the docstring from the main code.

I have created ticket [#7816](https://github.com/twisted/twisted/issues/7816) to look at opportunistic port of application.services


-------

Please check latest changes

with these changes I can do (build/py33/ is my py33 tox environment)

```
$ ./build/py33/bin/python3 bin/trial --version
Twisted version: 15.0.0
```

sending changes to buildbot...

don't know what test to run on buildbot for these  changes.
hope that soon we can have trial ported on py3 and integrate it with buildbots

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

Thanks Adi.

They look okay to me on first glance, but I'd prefer a second opinion (maybe Glyph's) to make sure that we're on the right track here -- as this isn't adding anything Trial to the distribution (which is what the policy says to do), but it's getting "bootstrappable" so that Trial can be used to actually test Trial.

I like this path, but, again, I defer.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

(In [44318]) Branching to py3-trial-version-7803-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

I have created new branch to fix conflicts and sent it to buildbot.
Please review the latest diff. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_8"></a>|@dustin commented|
|-|-|

I like this approach, too, not least because trial is a Buildbot requirement and it will be hard for us to tell whether Buildbot works on Python3 without Trial.

However, on the py3-trial-version-7803-2 branch (https://github.com/twisted/twisted/compare/py3-trial-version-7803-2), I'm running `python3 bin/trial twisted.application.test.test_service` (or for any test script) and seeing

```
  File "bin/trial", line 18, in <module>
    run()
  File "/home/dustin/code/twisted/t/twisted/twisted/scripts/trial.py", line 602, in run
    config.parseOptions()
  File "/home/dustin/code/twisted/t/twisted/twisted/python/usage.py", line 277, in parseOptions
    self.postOptions()
  File "/home/dustin/code/twisted/t/twisted/twisted/scripts/trial.py", line 473, in postOptions
    _BasicOptions.postOptions(self)
  File "/home/dustin/code/twisted/t/twisted/twisted/scripts/trial.py", line 383, in postOptions
    self['reporter'] = self._loadReporterByName(self['reporter'])
  File "/home/dustin/code/twisted/t/twisted/twisted/scripts/trial.py", line 370, in _loadReporterByName
    for p in plugin.getPlugins(itrial.IReporter):
  File "/home/dustin/code/twisted/t/twisted/twisted/plugin.py", line 209, in getPlugins
    allDropins = getCache(package)
  File "/home/dustin/code/twisted/t/twisted/twisted/plugin.py", line 143, in getCache
    for plugmod in mod.iterModules():
  File "/home/dustin/code/twisted/t/twisted/twisted/python/modules.py", line 134, in iterModules
    children.sort()
AttributeError: 'map' object has no attribute 'sort'
```

which I realize is a known issue (`children` now returns a map, not a list), but I'm not sure how this is causing me issues on this branch but hasn't bothered either of you.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [44679]) Branching to py3-trial-version-7803-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

Hi dustin, this is just a port for --version, to get the very basic part working, and then porting the actual trial bits :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

I've merged it forward and instead of altering spewer (which doesn't work on py3 with the changes made in this branch), I just make it so that spewer doesn't have to be imported if it's not being used, which sidesteps this a little.

Builders spun.

**Comment by twisted-trac** [comment]
|<img alt="adiroiban1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|adiroiban1 commented|
|-|-|

Thanks for working on this :)

What is the purpose of this comments ?

```
# Won't work on Python 3, because util.spewer is not ported.
```

with the current code, spew will fail at runtime so I don't think that we need to comments. 

I remember testing spewer on py3 and it was flooding my console.
I think that instead of leaving those comments we should create a separate ticket to port spewer and let people know that spew option does not work in python. The ticket can explain why spewer don't work.

Maybe the comment can be part of opt_spew docstring so that users will see that it is not yet working on py3.

Even if spewer is not yet working on Py3 maybe we can leave the change in utils so that the code can be compiled by py3 and we don't need to change the imports in trial.py

--------

Also I think that we should create a bug report from Dustin's  comment and have it as a blocker for porting trial on py3

Cheers!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

Hi hawkowl,

Thanks for working on this.

  1. The first thing I think that needs addressing here is the use of the word "opportunistic".  The dictionary definition I have is "exploiting chances offered by immediate circumstances without reference to a general plan or moral principle", which seems like not a great way to make changes :-).  I think this is my fault, because I might have suggested it when [ticket:7827 the ticket about twisted.persisted.styles] was filed to mean something like "incomplete port that allows a module to at least import".  Maybe we should just say "incomplete" or "partial" if that's all it means?  Maybe a better description would be "port just enough of `trial` to run the `--version` command line?"
  2. As adi rightly pointed out, the comments about un-ported code are not helpful; lots of code isn't ported and it isn't specifically called out in comments.  The links to specific tickets are helpful, so if you wanted to replace the general "won't work on py3" comments with links, that would also be fine.  We should just go ahead and port the code in further tickets and track what is and isn't ported by reference to tests, and start paying attention to coverage reports.  This does imply some additional work - once we have trial ported we should really get a py3 coverage builder going.  Can you file a ticket for that in an appropriate place? (Somewhere on twisted-infra probably?)
  3. In `test_service`, the additional `if not _PY3` assertion seems gross to me. Specifically, I don't think we should have tests which have different meanings on py3 and py2; if we have some behavior that isn't yet ported, we should have a separate test in a separate module that is specifically not marked as ported yet so that it will be caught by the normal process.  By leaving an assertion in a test like this, following the normal process, we may end up with a bunch of edge-cases that are not covered on py3.  I think that if the tests were split into py2 and py3 modules that covered both halves of this then the similarly-problematic conditional in the implementation would be adequately covered and someone would quickly notice it needed to be ported.
  4. Speaking of `test_service`, "`sut`" is not a very meaningful variable name.  I am guessing you mean "system-under-test" but in this case it's actually the _unit_ under test, the system under test would be the whole module or maybe even Twisted as a whole :).  So let's go with a more meaningful name like "app" or somesuch.

Speaking of that [ticket:7827 persisted ticket I never finished], I guess I should just finish that so that fewer things should have weird edge cases sticking out that need stuff like that `if _PY3` conditional.  So again, my fault :).

Please address these points as best you can and submit for re-review.  I would say to just land but I think we need to be really careful about the test coverage discipline we're using for saying py3 is "supported".

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl removed owner|
|-|-|

Fixed up a few things according to the comments. Builders spun, look green so far, please review.

**Comment by twisted-trac** [comment]
|<img alt="adiroiban1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|adiroiban1 commented|
|-|-|

Changes look good!

I think that comments like this should stay and be updated to the new style guide

https://github.com/twisted/twisted/blob/trunk/docs/core/development/policy/coding-standard.rst#comments

```
# TODO https://twistedmatrix.com/trac/ticket/6910
```

I wrote the code before the FIXME convention was accepted in trunk.

I think that all references to tickets should use the same standard like

```
# FIXME: https://twistedmatrix.com/trac/ticket/6910
```

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Regarding comment 8 , this should be already fixed by [#7804](https://github.com/twisted/twisted/issues/7804)

------------

Please update the FIXME comments and merge. Thanks!

Looking forward for running trial under py3 :)

------------

Just a minor comment

I like `sut` as it can also stand for `stuff under test` :)

I am using this as a standard way to highlight the object which is tested for unit/system/integration/black-back/functional tests. From the context of the test case it should be easier to see that the sut is a protocol/factory/application/logger/observer/http client/http server ...etc

The idea is that in a test you might have many factories, protocols, clients, servers and other object and when scanning old cold is hard to see if the `f` variable is a dependent factory or the factory under test.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban commented|
|-|-|

FIXME are ok. Sorry for the previous comment.

Please merge! Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44885]) Merge py3-trial-version-7803-3: Port just enough of trial to run --version on Python 3

Author: adiroiban, hawkowl
Reviewers: hawkowl, glyph, adiroiban
Fixes: [#7803](https://github.com/twisted/twisted/issues/7803)
